### ISSUE TITLE
Object field caching

### ISSUE THREAD

**Issue by moleike** [moleike/haskell-jsonnet#30]
As currently implemented, mixins are extremely slow (and possibly leak), take this benchmark for example:

```jsonnet
  local fibnext = {
    a: super.a + super.b,
    b: super.a,
  };
  local fib(n) =
    if n == 0 then
      { a: 1, b: 1 }
    else
      fib(n - 1) + fibnext;

  fib(25)
```

This is runs in exponential time like a naive recursive implementation because `super` doesn't allow us to use memoization, so we end up recomputing every Fibonacci number.

This seems has been addressed in [go-jsonnet](https://github.com/google/go-jsonnet/pull/280) and sjsonnet also supports some form of [caching](https://github.com/databricks/sjsonnet/blob/master/sjsonnet/src/sjsonnet/Val.scala#L90).

**Comment by moleike** [comment]
With the following experimental evaluator, we can run this benchmark in polynomial time, by memoising the result of expressions that contain super (but no self): https://gist.github.com/moleike/f50931f54b162e3644da27d6f50e8c18

**Comment by sbarzowski** [comment]
FWIW the way caching works in go-jsonnet is that we cache field values by (name, depth) where depth is how deep in the inheritance field is (0 is the top, 1 is one super level deep, 2 is two super levels deep etc.). Inheritance is associative so we can always look at the object as a stack of layers, with each field at each level well-defined.

The cache is only for a specific object, so if you override any field (add more layers) it's considered a completely new object with a clean cache.

**Comment by moleike** [comment]
@sbarzowski thanks for sharing how it's done in `go-jsonnet`, I definitely will look there in the future for inspiration. I am doing something different to avoid recomputing now.

**Comment by sbarzowski** [comment]
There are two main resources:
1) https://github.com/google/go-jsonnet/blob/v0.17.0/value.go#L420 <- Here's the definition and basic operations. It is relatively straightforward and well-commented. If you have any questions, ask away.
2) https://jsonnet.org/ref/language.html#object explains my mental model of objects in Jsonnet (especially the stack of layers part).

**Comment by moleike** [comment]
Fixed in 6fc7be1dabd9b43070a08a05f2424e23c4506eeb
